From: Simon McVittie <smcv@collabora.com>
Date: Thu, 9 Apr 2026 20:06:18 +0100
Subject: utils: Move flatpak_get_path_for_fd to here

This was originally in flatpak-portal, then was duplicated into
flatpak-run in commit ac62ebe3 "run: Use O_PATH fds for the runtime and
app deploy directories", and subsequently removed from the portal in
commit 3c500145 "portal: Use --bind-fd, --app-fd and --usr-fd options to
avoid races". Now we want to use it in the portal again.

Co-authored-by: Sebastian Wick <sebastian.wick@redhat.com>
Signed-off-by: Simon McVittie <smcv@collabora.com>
Bug: https://github.com/flatpak/flatpak/issues/6584
Forwarded: https://github.com/flatpak/flatpak/pull/6588
---
 common/flatpak-run.c           | 44 ++++---------------------------------
 common/flatpak-utils-private.h |  3 +++
 common/flatpak-utils.c         | 50 ++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 57 insertions(+), 40 deletions(-)

diff --git a/common/flatpak-run.c b/common/flatpak-run.c
index f176e88..7d46dd6 100644
--- a/common/flatpak-run.c
+++ b/common/flatpak-run.c
@@ -2939,42 +2939,6 @@ open_namespace_fd_if_needed (const char *path,
   return -1;
 }
 
-static char *
-get_path_for_fd (int        fd,
-                 GError   **error)
-{
-  g_autofree char *proc_path = NULL;
-  g_autofree char *path = NULL;
-
-  proc_path = g_strdup_printf ("/proc/self/fd/%d", fd);
-  path = glnx_readlinkat_malloc (AT_FDCWD, proc_path, NULL, error);
-  if (path == NULL)
-    return NULL;
-
-  /* All normal paths start with /, but some weird things
-     don't, such as socket:[27345] or anon_inode:[eventfd].
-     We don't support any of these */
-  if (path[0] != '/')
-    {
-      return glnx_null_throw (error, "%s resolves to non-absolute path %s",
-                              proc_path, path);
-    }
-
-  /* File descriptors to actually deleted files have " (deleted)"
-     appended to them. This also happens to some fake fd types
-     like shmem which are "/<name> (deleted)". All such
-     files are considered invalid. Unfortunately this also
-     matches files with filenames that actually end in " (deleted)",
-     but there is not much to do about this. */
-  if (g_str_has_suffix (path, " (deleted)"))
-    {
-      return glnx_null_throw (error, "%s resolves to deleted path %s",
-                              proc_path, path);
-    }
-
-  return g_steal_pointer (&path);
-}
-
 gboolean
 flatpak_run_app (FlatpakDecomposed   *app_ref,
                  FlatpakDeploy       *app_deploy,
@@ -3203,7 +3167,7 @@ flatpak_run_app (FlatpakDecomposed   *app_ref,
     {
       g_autofree char *path = NULL;
 
-      path = get_path_for_fd (custom_runtime_fd, &my_error);
+      path = flatpak_get_path_for_fd (custom_runtime_fd, &my_error);
       if (path == NULL)
         {
           return flatpak_fail_error (error, FLATPAK_ERROR,
@@ -3331,7 +3295,7 @@ flatpak_run_app (FlatpakDecomposed   *app_ref,
     {
       g_autofree char *path = NULL;
 
-      path = get_path_for_fd (custom_app_fd, error);
+      path = flatpak_get_path_for_fd (custom_app_fd, error);
       if (path == NULL)
         return glnx_prefix_error (error, "Cannot convert custom app fd to path");
 
@@ -3640,7 +3604,7 @@ flatpak_run_app (FlatpakDecomposed   *app_ref,
 
       /* We get the path the fd refers to, to determine to mount point
        * destination inside the sandbox */
-      path = get_path_for_fd (fd, error);
+      path = flatpak_get_path_for_fd (fd, error);
       if (!path)
         return FALSE;
 
@@ -3657,7 +3621,7 @@ flatpak_run_app (FlatpakDecomposed   *app_ref,
 
       /* We get the path the fd refers to, to determine to mount point
        * destination inside the sandbox */
-      path = get_path_for_fd (fd, error);
+      path = flatpak_get_path_for_fd (fd, error);
       if (!path)
         return FALSE;
 
diff --git a/common/flatpak-utils-private.h b/common/flatpak-utils-private.h
index e7fa348..8cad667 100644
--- a/common/flatpak-utils-private.h
+++ b/common/flatpak-utils-private.h
@@ -351,6 +351,9 @@ gboolean flatpak_is_debugging (void);
 int flatpak_parse_fd (const char  *fd_string,
                       GError     **error);
 
+char * flatpak_get_path_for_fd (int      fd,
+                                GError **error);
+
 #define FLATPAK_MESSAGE_ID "c7b39b1e006b464599465e105b361485"
 
 gboolean flatpak_set_cloexec (int fd);
diff --git a/common/flatpak-utils.c b/common/flatpak-utils.c
index 706a9cb..b8bfcfd 100644
--- a/common/flatpak-utils.c
+++ b/common/flatpak-utils.c
@@ -2542,3 +2542,53 @@ flatpak_set_cloexec (int fd)
 
   return TRUE;
 }
+
+/*
+ * Attempt to discover the filesystem path corresponding to @fd.
+ *
+ * If @fd points to an existing file, return the absolute path of that
+ * file in the environment where it was opened. Note that this is not
+ * necessarily a valid path in the current namespace, if it was
+ * transferred via fd-passing from a process in a different filesystem
+ * namespace.
+ *
+ * If @fd points to a deleted file, or to a socket, fifo, memfd or similar
+ * non-filesystem object, set an error and return %NULL.
+ *
+ * Returns: (type filename) (transfer full) (nullable):
+ */
+char *
+flatpak_get_path_for_fd (int        fd,
+                         GError   **error)
+{
+  g_autofree char *proc_path = NULL;
+  g_autofree char *path = NULL;
+
+  proc_path = g_strdup_printf ("/proc/self/fd/%d", fd);
+  path = glnx_readlinkat_malloc (AT_FDCWD, proc_path, NULL, error);
+  if (path == NULL)
+    return NULL;
+
+  /* All normal paths start with /, but some weird things
+     don't, such as socket:[27345] or anon_inode:[eventfd].
+     We don't support any of these */
+  if (path[0] != '/')
+    {
+      return glnx_null_throw (error, "%s resolves to non-absolute path %s",
+                              proc_path, path);
+    }
+
+  /* File descriptors to actually deleted files have " (deleted)"
+     appended to them. This also happens to some fake fd types
+     like shmem which are "/<name> (deleted)". All such
+     files are considered invalid. Unfortunately this also
+     matches files with filenames that actually end in " (deleted)",
+     but there is not much to do about this. */
+  if (g_str_has_suffix (path, " (deleted)"))
+    {
+      return glnx_null_throw (error, "%s resolves to deleted path %s",
+                              proc_path, path);
+    }
+
+  return g_steal_pointer (&path);
+}
